types/ember/index.d.ts

```diff
@@ -836,7 +850,7 @@ declare module 'ember' {
             static create<Instance>(this: EmberClassConstructor<Instance>): Fix<Instance>;
 
             static create<Instance, Args, T1 extends EmberInstanceArguments<Args>>(
-                this: EmberClassConstructor<Instance & ComputedPropertyGetters<Args>>,
+                this: EmberClassConstructor<UnwrapComputedPropertyGetters<Args> & Instance>,
                 arg1: T1 & ThisType<Fix<T1 & Instance>>
```

Comment:
With TS 2.8 I think we can declare it more simply:
```ts
static create<Class extends typeof Ember.CoreObject, Arg1>(
    this: Class,
    arg1: Arg1 & ThisType<Arg1 & InstanceType<Class>>
): InstanceType<Class> & Arg1;
```

I'll call out a couple of things that are missing compared to the original declaration.

- `EmberInstanceArguments<T>` verified that the arguments passed to `create()` matched the types on the class. E.g. `Object.extend({ x: 42 }).create({ x: 'foo' })` would not typecheck (although it does work at runtime-- `x` is a string).

- `Fix<T>` worked around https://github.com/typed-ember/ember-typings/pull/32. I suspect it's no longer needed.

Reply:
I think it would be useful for us to take a pass at `create` and the associated code as a follow-on effort and see what we can clean up there (as well as fleshing out the test coverage).

In the meantime, though, hopefully there's a low-lift way to keep things working roughly as-is with `UnwrapComputedPropertyGetters` to get this PR landed?

Reply:
In the future I also want us to stop `&`ing `InstanceType<Class> & Arg1`. I don't believe collisions are handled properly.
```ts
{ // InstanceType<Class>
   foo: string
} & { // .create args
   foo: number
} = {
  foo: string & number
}

it should be 

{
   foo: number
}

```

Maybe something like

```ts
type Mix<A, B> = B & Pick<A, Exclude<keyof A, keyof B>>;

```
[example](https://www.typescriptlang.org/play/#src=type%20Mix%3CA%2C%20B%3E%20%3D%20B%20%26%20Pick%3CA%2C%20Exclude%3Ckeyof%20A%2C%20keyof%20B%3E%3E%3B%0D%0A%0D%0Atype%20AA%20%3D%20%7B%0D%0A%20%20%20%20foo%3A%20string%3B%0D%0A%20%20%20%20bar%3A%20number%3B%0D%0A%20%20%20%20baz%3A%20boolean%0D%0A%7D%0D%0A%0D%0Atype%20BB%20%7B%0D%0A%20%20%20%20a%3A%20string%3B%0D%0A%20%20%20%20b%3A%20number%3B%0D%0A%20%20%20%20baz%3A%20string%5B%5D%3B%0D%0A%7D%0D%0A%0D%0Atype%20CC%20%3D%20Mix%3CAA%2C%20BB%3E%3B%0D%0Aconst%20x%3A%20CC%3B)

Reply:
Agreed that `&`ing together the types is incorrect there (and doubtless there are other places we're doing that as well that should be fixed), but shouldn't code like this be a type error regardless?

```ts
const MyClass = EmberObject.extend({
  foo: 'hello'
});
MyClass.create({ foo: 5 });
```

Seems like (roughly) the moral equivalent of

```ts
class MyClass {
  constructor(public foo: string = 'hello') {}
}
new MyClass(5);
```

Reply:
I need someone who has more context to look at the `.create` stuff. I've spent a few hours trying to guess what the various `Args`, `T1`, etc... are. As far as I can tell, none of the suggestions in this thread would allow the negative test cases to appropriately error while keeping the positive test cases passing.

###### create-negative.ts
```ts
import { assertType } from './lib/assert';
import Ember from 'ember';
import { PersonWithNumberName, Person } from './create';

const p3 = Person.create({ firstName: 99 }); // $ExpectError
const p2b = Person.create({}, { firstName: 99 }); // $ExpectError
const p2c = Person.create({}, {}, { firstName: 99 }); // $ExpectError

const p4 = new PersonWithNumberName();
// assertType<number>(p4.get('fullName'));

assertType<Ember.ComputedProperty<string, string>>(p4.fullName); // $ExpectError

```
###### create.ts
```ts
import Ember from 'ember';
import { assertType } from './lib/assert';

const o = Ember.Object.create();
assertType<object>(o);

const o1 = Ember.Object.create({x: 9});
assertType<number>(o1.x);

const obj = Ember.Object.create({ a: 1 }, { b: 2 }, { c: 3 });
assertType<number>(obj.b);
assertType<number>(obj.a);
assertType<number>(obj.c);

export class Person extends Ember.Object.extend({
    fullName: Ember.computed('firstName', 'lastName', function() {
        return [this.firstName + this.lastName].join(' ');
    })
}) {
    firstName: string;
    lastName: string;
    age: number;
}
const p = new Person();
assertType<string>(p.firstName);
assertType<Ember.ComputedProperty<string>>(p.fullName);
assertType<string>(p.get('fullName'));

const p2 = Person.create({ firstName: 'string' });
const p2b = Person.create({}, { firstName: 'string' });
const p2c = Person.create({}, {}, { firstName: 'string' });

export class PersonWithNumberName extends Person.extend({
  fullName: 6
}) {}

const p4 = new PersonWithNumberName();
assertType<string>(p4.firstName);
assertType<number>(p4.fullName);
```

Reply:
@mike-north I'm not positive I'm in a better position than you are w.r.t. the current mechanics of `create`, but I'm happy to take a pass later today and see if I can get the tests passing if @dwickern doesn't beat me to it.

Reply:
Unless it complicates the typing significantly, I believe it should, as long as the `Set` type is compatible. It's a little weird in this specific example since the computed isn't actually set up to work with `set`, but that's not currently something we try to capture in the type system.

Reply:
I'll give you both commit access to my fork. Unfortunately I'm out of time to spend on this for a few days

Reply:
The `Ember.Object.create` issue is now resolved. Positive and negative test cases added to reflect this for 0-4 arguments passed to `.create`